File: src/env/template.ts
const TOKEN_RE = /\$\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g;

export function findPortReferences(template: string): string[] {
  return Array.from(template.matchAll(TOKEN_RE), (m) => m[1]);
}

export function renderEnv(
  env: Record<string, string>,
  ports: Record<string, number>,
): Record<string, string> {
  const result: Record<string, string> = {};
  for (const [key, template] of Object.entries(env)) {
    result[key] = template.replace(TOKEN_RE, (_match, name) => {
      const port = ports[name];
      if (port === undefined) {
        throw new Error(`env.${key} references unknown port '${name}'`);
      }
      return String(port);
    });
  }
  return result;
}

export function renderEnvOrExit(
  env: Record<string, string>,
  ports: Record<string, number>,
): Record<string, string> {
  try {
    return renderEnv(env, ports);
  } catch (err) {
    console.error(`Error: ${(err as Error).message}`);
    process.exit(1);
  }
}

File: internal/envfile/template.go
package envfile

import (
	"fmt"
	"regexp"
)

var tokenRE = regexp.MustCompile(`\$\{([a-zA-Z_][a-zA-Z0-9_]*)\}`)

// FindPortReferences returns all port names referenced as ${NAME} in template,
// in order of appearance, with duplicates preserved.
func FindPortReferences(template string) []string {
	matches := tokenRE.FindAllStringSubmatch(template, -1)
	out := make([]string, 0, len(matches))
	for _, m := range matches {
		out = append(out, m[1])
	}
	return out
}

// Render expands ${NAME} references in each template using ports.
// Returns an error referencing the env-var key if any template names an
// unknown port.
func Render(env map[string]string, ports map[string]int) (map[string]string, error) {
	out := make(map[string]string, len(env))
	for key, template := range env {
		var renderErr error
		result := tokenRE.ReplaceAllStringFunc(template, func(match string) string {
			name := match[2 : len(match)-1] // strip ${ and }
			port, ok := ports[name]
			if !ok {
				if renderErr == nil {
					renderErr = fmt.Errorf("env.%s references unknown port '%s'", key, name)
				}
				return match
			}
			return fmt.Sprintf("%d", port)
		})
		if renderErr != nil {
			return nil, renderErr
		}
		out[key] = result
	}
	return out, nil
}
